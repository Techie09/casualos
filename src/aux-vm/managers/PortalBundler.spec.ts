import {
    createBot,
    createPrecalculatedBot,
    OpenCustomPortalOptions,
    stateUpdatedEvent,
} from '@casual-simulation/aux-common';
import { waitAsync } from '@casual-simulation/aux-common/test/TestHelpers';
import { Subscription } from 'rxjs';
import {
    DEFAULT_BASE_MODULE_URL,
    ESBuildPortalBundler,
    PortalBundler,
    ScriptPrefix,
} from './PortalBundler';

console.log = jest.fn();
jest.mock('axios');

describe('ESBuildPortalBundler', () => {
    let bundler: PortalBundler;

    describe('bundleTag()', () => {
        let func1: jest.Mock<any>;
        let func2: jest.Mock<any>;

        beforeEach(() => {
            bundler = new ESBuildPortalBundler();
            (<any>globalThis).func1 = func1 = jest.fn();
            (<any>globalThis).func2 = func2 = jest.fn();
        });

        afterEach(() => {
            delete (<any>globalThis).func2;
            delete (<any>globalThis).func1;
        });

        const prefixCases: [string, string[], string, string][] = [
            ['library emoji', ['ðŸ“–'], 'ðŸ“–', 'ðŸ“–'],
            ['smile emoji', ['ðŸ™‚'], 'ðŸ™‚', 'ðŸ™‚'],
            ['multiple emoji', ['ðŸ“–', 'ðŸ™‚'], 'ðŸ™‚', 'ðŸ“–'],
        ];

        describe.each(prefixCases)(
            '%s',
            (desc, scriptPrefixes, firstPrefix, secondPrefix) => {
                let prefixes: ScriptPrefix[];

                beforeEach(() => {
                    prefixes = [
                        {
                            prefix: firstPrefix,
                            language: 'javascript',
                        },
                        {
                            prefix: secondPrefix,
                            language: 'javascript',
                        },
                    ];
                });

                it('should resolve with null if there are no tags with the right prefix', async () => {
                    const state = {
                        bot1: createPrecalculatedBot('bot1', {
                            main: `console.log("abc")`,
                        }),
                        bot2: createPrecalculatedBot('bot2', {
                            main: `console.log("def")`,
                        }),
                    };
                    const bundle = await bundler.bundleTag(
                        state,
                        'main',
                        prefixes
                    );

                    expect(bundle).toEqual(null);
                });

                it('should resolve with a bundle that contains the specified tags', async () => {
                    const state = {
                        bot1: createPrecalculatedBot('bot1', {
                            main: `${firstPrefix}console.log("abc")`,
                        }),
                        bot2: createPrecalculatedBot('bot2', {
                            main: `${secondPrefix}console.log("def")`,
                        }),
                    };

                    const bundle = await bundler.bundleTag(
                        state,
                        'main',
                        prefixes
                    );

                    expect(bundle).not.toEqual(null);
                    expect(bundle.source).toBeTruthy();
                    expect(bundle).toMatchSnapshot();
                });

                it('should resolve with build errors', async () => {
                    const state = {
                        bot1: createPrecalculatedBot('bot1', {
                            main: `${firstPrefix}console.log("ab`,
                        }),
                    };

                    const bundle = await bundler.bundleTag(
                        state,
                        'main',
                        prefixes
                    );

                    expect(bundle).not.toEqual(null);
                    expect(bundle.error).toBeTruthy();
                    expect(bundle).toMatchSnapshot();
                });

                it('should execute entry points in bot ID alphabetical order', async () => {
                    const state = {
                        def: createPrecalculatedBot('def', {
                            main: `${firstPrefix}globalThis.func1("second");`,
                        }),
                        abc: createPrecalculatedBot('abc', {
                            main: `${secondPrefix}globalThis.func1("first");`,
                        }),
                    };

                    const bundle = await bundler.bundleTag(
                        state,
                        'main',
                        prefixes
                    );

                    expect(bundle).not.toEqual(null);
                    expect(bundle.source).toBeTruthy();

                    eval(bundle.source);

                    expect(func1).toBeCalledTimes(2);
                    expect(func1).toHaveBeenNthCalledWith(1, 'first');
                    expect(func1).toHaveBeenNthCalledWith(2, 'second');
                });

                it('should be able to import scripts from other tags', async () => {
                    const state = {
                        abc: createPrecalculatedBot('abc', {
                            main: `${firstPrefix}import "${secondPrefix}other"; globalThis.func1("main");`,
                            other: `${secondPrefix}globalThis.func1("other");`,
                        }),
                    };

                    const bundle = await bundler.bundleTag(
                        state,
                        'main',
                        prefixes
                    );

                    expect(bundle).not.toEqual(null);
                    expect(bundle.source).toBeTruthy();

                    eval(bundle.source);

                    expect(func1).toBeCalledTimes(2);
                    expect(func1).toHaveBeenNthCalledWith(1, 'other');
                    expect(func1).toHaveBeenNthCalledWith(2, 'main');
                });

                it('should handle modules that reference each other', async () => {
                    const state = {
                        abc: createPrecalculatedBot('abc', {
                            main: `${firstPrefix}import "${secondPrefix}other"; globalThis.func1("main");`,
                            other: `${secondPrefix}import "${firstPrefix}main"; globalThis.func1("other");`,
                        }),
                    };

                    const bundle = await bundler.bundleTag(
                        state,
                        'main',
                        prefixes
                    );

                    expect(bundle).not.toEqual(null);
                    expect(bundle.source).toBeTruthy();

                    eval(bundle.source);

                    expect(func1).toBeCalledTimes(2);
                    expect(func1).toHaveBeenNthCalledWith(1, 'other');
                    expect(func1).toHaveBeenNthCalledWith(2, 'main');
                });

                it('should report which bots and tags are included in the bundle', async () => {
                    const state = {
                        bot1: createPrecalculatedBot('bot1', {
                            main: `${firstPrefix}console.log("abc")`,
                            different: true,
                        }),
                        bot2: createPrecalculatedBot('bot2', {
                            main: `${secondPrefix}import "${secondPrefix}second"; console.log("def")`,
                            second: `${secondPrefix}let test = 123;`,
                        }),
                        bot3: createPrecalculatedBot('bot3', {
                            main: `no prefix`,
                        }),
                        bot4: createPrecalculatedBot('bot4', {
                            other: `no prefix`,
                        }),
                    };

                    const bundle = await bundler.bundleTag(
                        state,
                        'main',
                        prefixes
                    );

                    expect(bundle).not.toEqual(null);
                    expect(bundle.modules).toEqual({
                        bot1: new Set(['main']),
                        bot2: new Set(['main', 'second']),
                    });
                });

                describe('imports', () => {
                    beforeEach(() => {
                        require('axios').__reset();
                    });

                    it('should try to load modules from skypack', async () => {
                        require('axios').__setResponse({
                            data: `export const fun = globalThis.func1;`,
                        });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        await waitAsync();

                        expect(bundle).not.toEqual(null);
                        expect(bundle.source).toBeTruthy();
                        let [url] = require('axios').__getLastGet();

                        expect(url).toBe(
                            `${DEFAULT_BASE_MODULE_URL}/lodash?dts`
                        );

                        eval(bundle.source);

                        expect(func1).toBeCalledTimes(1);
                    });

                    it('should report errors that occur while fetching data', async () => {
                        require('axios').__setFail(true);

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        expect(bundle).not.toEqual(null);
                        expect(bundle.error).toBeTruthy();
                        expect(bundle.error).toMatchSnapshot();
                    });

                    it('should support HTTPS modules that have relative references', async () => {
                        require('axios')
                            .__setNextResponse({
                                data: `export * from './fun';`,
                            })
                            .__setNextResponse({
                                data: `export const fun = globalThis.func1;`,
                            });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        expect(bundle).not.toEqual(null);
                        expect(bundle.source).toBeTruthy();
                        let requests = require('axios').__getRequests();

                        expect(requests).toEqual([
                            ['get', `${DEFAULT_BASE_MODULE_URL}/lodash?dts`],
                            ['get', `${DEFAULT_BASE_MODULE_URL}/lodash/fun`],
                        ]);

                        eval(bundle.source);

                        expect(func1).toBeCalledTimes(1);
                    });

                    it('should support HTTPS modules that have parent references', async () => {
                        require('axios')
                            .__setNextResponse({
                                data: `export * from './first.js';`,
                            })
                            .__setNextResponse({
                                data: `export * from '../nested/fun.js';`,
                            })
                            .__setNextResponse({
                                data: `export const fun = globalThis.func1;`,
                            });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash/test/haha.js"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        expect(bundle).not.toEqual(null);
                        expect(bundle.source).toBeTruthy();
                        let requests = require('axios').__getRequests();

                        expect(requests).toEqual([
                            [
                                'get',
                                `${DEFAULT_BASE_MODULE_URL}/lodash/test/haha.js?dts`,
                            ],
                            [
                                'get',
                                `${DEFAULT_BASE_MODULE_URL}/lodash/test/first.js`,
                            ],
                            [
                                'get',
                                `${DEFAULT_BASE_MODULE_URL}/lodash/nested/fun.js`,
                            ],
                        ]);

                        eval(bundle.source);

                        expect(func1).toBeCalledTimes(1);
                    });

                    it('should support HTTPS modules that have deep parent references', async () => {
                        require('axios')
                            .__setNextResponse({
                                data: `export * from './first.js';`,
                            })
                            .__setNextResponse({
                                data: `export * from '../../nested/fun.js';`,
                            })
                            .__setNextResponse({
                                data: `export const fun = globalThis.func1;`,
                            });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash/test/haha.js"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        expect(bundle).not.toEqual(null);
                        expect(bundle.source).toBeTruthy();
                        let requests = require('axios').__getRequests();

                        expect(requests).toEqual([
                            [
                                'get',
                                `${DEFAULT_BASE_MODULE_URL}/lodash/test/haha.js?dts`,
                            ],
                            [
                                'get',
                                `${DEFAULT_BASE_MODULE_URL}/lodash/test/first.js`,
                            ],
                            ['get', `${DEFAULT_BASE_MODULE_URL}/nested/fun.js`],
                        ]);

                        eval(bundle.source);

                        expect(func1).toBeCalledTimes(1);
                    });

                    it('should support HTTPS modules that have absolute references', async () => {
                        require('axios')
                            .__setNextResponse({
                                data: `export * from '/fun';`,
                            })
                            .__setNextResponse({
                                data: `export const fun = globalThis.func1;`,
                            });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        expect(bundle).not.toEqual(null);
                        expect(bundle.source).toBeTruthy();

                        let requests = require('axios').__getRequests();

                        expect(requests).toEqual([
                            ['get', `${DEFAULT_BASE_MODULE_URL}/lodash?dts`],
                            ['get', `${DEFAULT_BASE_MODULE_URL}/fun`],
                        ]);

                        eval(bundle.source);

                        expect(func1).toBeCalledTimes(1);
                    });

                    it('should support HTTPS modules that have nested references', async () => {
                        require('axios')
                            .__setNextResponse({
                                data: `export * from './fun';`,
                            })
                            .__setNextResponse({
                                data: `export * from './other';`,
                            })
                            .__setNextResponse({
                                data: `export * from '/final';`,
                            })
                            .__setNextResponse({
                                data: `export const fun = globalThis.func1;`,
                            });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        expect(bundle).not.toEqual(null);
                        expect(bundle.source).toBeTruthy();

                        let requests = require('axios').__getRequests();

                        expect(requests).toEqual([
                            ['get', `${DEFAULT_BASE_MODULE_URL}/lodash?dts`],
                            ['get', `${DEFAULT_BASE_MODULE_URL}/lodash/fun`],
                            [
                                'get',
                                `${DEFAULT_BASE_MODULE_URL}/lodash/fun/other`,
                            ],
                            ['get', `${DEFAULT_BASE_MODULE_URL}/final`],
                        ]);

                        eval(bundle.source);

                        expect(func1).toBeCalledTimes(1);
                    });

                    it('should cache HTTP modules across builds', async () => {
                        require('axios')
                            .__setNextResponse({
                                data: `export const fun = globalThis.func1;`,
                            })
                            .__setNextResponse({
                                data: `export const fun = globalThis.func2;`,
                            });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const bundle1 = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );
                        const bundle2 = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        let requests = require('axios').__getRequests();

                        expect(requests).toEqual([
                            ['get', `${DEFAULT_BASE_MODULE_URL}/lodash?dts`],
                        ]);
                        expect(bundle1).toEqual(bundle2);
                    });

                    it('should cache HTTP modules that are requested concurrently', async () => {
                        require('axios')
                            .__setNextResponse({
                                data: `export const fun = globalThis.func1;`,
                            })
                            .__setNextResponse({
                                data: `export const fun = globalThis.func2;`,
                            });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const [bundle1, bundle2] = await Promise.all([
                            bundler.bundleTag(state, 'main', prefixes),
                            bundler.bundleTag(state, 'main', prefixes),
                        ]);

                        let requests = require('axios').__getRequests();

                        expect(requests).toEqual([
                            ['get', `${DEFAULT_BASE_MODULE_URL}/lodash?dts`],
                        ]);
                        expect(bundle1).toEqual(bundle2);
                    });

                    it('should report external modules that were imported', async () => {
                        require('axios').__setResponse({
                            data: `export const fun = globalThis.func1;`,
                        });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        await waitAsync();

                        expect(bundle).not.toEqual(null);
                        expect(bundle.externals).toEqual({
                            lodash: {
                                id: 'lodash',
                                url: `${DEFAULT_BASE_MODULE_URL}/lodash?dts`,
                                typescriptDefinitionsURL: null,
                            },
                        });
                    });

                    it('should report typescript definitions that were returned from the server', async () => {
                        require('axios').__setResponse({
                            data: `export const fun = globalThis.func1;`,
                            headers: {
                                'x-typescript-types': '/typescriptDefinitions',
                            },
                        });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const bundle = await bundler.bundleTag(
                            state,
                            'main',
                            prefixes
                        );

                        await waitAsync();

                        expect(bundle).not.toEqual(null);
                        expect(bundle.externals).toEqual({
                            lodash: {
                                id: 'lodash',
                                url: `${DEFAULT_BASE_MODULE_URL}/lodash?dts`,
                                typescriptDefinitionsURL: `${DEFAULT_BASE_MODULE_URL}/typescriptDefinitions`,
                            },
                        });
                    });

                    it('should report typescript definitions for cached modules', async () => {
                        require('axios')
                            .__setNextResponse({
                                data: `export const fun = globalThis.func1;`,
                                headers: {
                                    'x-typescript-types':
                                        '/typescriptDefinitions',
                                },
                            })
                            .__setNextResponse({
                                data: `export const fun = globalThis.func2;`,
                            });

                        const state = {
                            bot1: createPrecalculatedBot('bot1', {
                                main: `${firstPrefix}import { fun } from "lodash"; fun();`,
                            }),
                        };

                        const [bundle1, bundle2] = await Promise.all([
                            bundler.bundleTag(state, 'main', prefixes),
                            bundler.bundleTag(state, 'main', prefixes),
                        ]);

                        let requests = require('axios').__getRequests();

                        expect(requests).toEqual([
                            ['get', `${DEFAULT_BASE_MODULE_URL}/lodash?dts`],
                        ]);
                        expect(bundle1.externals).toEqual({
                            lodash: {
                                id: 'lodash',
                                url: `${DEFAULT_BASE_MODULE_URL}/lodash?dts`,
                                typescriptDefinitionsURL: `${DEFAULT_BASE_MODULE_URL}/typescriptDefinitions`,
                            },
                        });
                        expect(bundle1).toEqual(bundle2);
                    });
                });
            }
        );

        it('should support typescript', async () => {
            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ“–let abc: string = "Hello!";`,
                }),
            };

            const bundle = await bundler.bundleTag(state, 'main', [
                {
                    prefix: 'ðŸ“–',
                    language: 'typescript',
                },
            ]);

            expect(bundle).not.toEqual(null);
            expect(bundle.source).toBeTruthy();
            expect(bundle).toMatchSnapshot();
        });

        it('should support JSON', async () => {
            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ“–{ "abc": "def" }`,
                }),
            };

            const bundle = await bundler.bundleTag(state, 'main', [
                {
                    prefix: 'ðŸ“–',
                    language: 'json',
                },
            ]);

            expect(bundle).not.toEqual(null);
            expect(bundle.source).toBeTruthy();
            expect(bundle).toMatchSnapshot();
        });

        it('should support JSX', async () => {
            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ“–let element = (<h1>Hello!</h1>);`,
                }),
            };

            const bundle = await bundler.bundleTag(state, 'main', [
                {
                    prefix: 'ðŸ“–',
                    language: 'jsx',
                },
            ]);

            expect(bundle).not.toEqual(null);
            expect(bundle.source).toBeTruthy();
            expect(bundle).toMatchSnapshot();
        });

        it('should support TSX', async () => {
            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ“–let element: any = (<h1>Hello!</h1>);`,
                }),
            };

            const bundle = await bundler.bundleTag(state, 'main', [
                {
                    prefix: 'ðŸ“–',
                    language: 'tsx',
                },
            ]);

            expect(bundle).not.toEqual(null);
            expect(bundle.source).toBeTruthy();
            expect(bundle).toMatchSnapshot();
        });

        it('should support prefixes in the given tag name', async () => {
            const prefixes: ScriptPrefix[] = [
                {
                    prefix: 'ðŸ”º',
                    language: 'javascript',
                },
                {
                    prefix: 'ðŸ“–',
                    language: 'javascript',
                },
            ];
            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ”ºglobalThis.func1("first");`,
                }),
                bot2: createPrecalculatedBot('bot2', {
                    main: `ðŸ“–globalThis.func1("second");`,
                }),
            };

            const bundle = await bundler.bundleTag(state, `ðŸ”ºmain`, prefixes);

            expect(bundle).not.toEqual(null);
            expect(bundle.source).toBeTruthy();

            eval(bundle.source);

            expect(func1).toBeCalledTimes(1);
            expect(func1).toBeCalledWith('first');
        });

        it('should support importing separate prefixes from special entry prefixes', async () => {
            const prefixes: ScriptPrefix[] = [
                {
                    prefix: 'ðŸ”º',
                    language: 'javascript',
                },
                {
                    prefix: 'ðŸ“–',
                    language: 'javascript',
                },
            ];
            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ”ºimport "ðŸ“–main"; globalThis.func1("first");`,
                }),
                bot2: createPrecalculatedBot('bot2', {
                    main: `ðŸ“–globalThis.func1("second");`,
                }),
            };

            const bundle = await bundler.bundleTag(state, `ðŸ”ºmain`, prefixes);

            expect(bundle).not.toEqual(null);
            expect(bundle.source).toBeTruthy();

            eval(bundle.source);

            expect(func1).toBeCalledTimes(2);
            expect(func1).toHaveBeenNthCalledWith(1, 'second');
            expect(func1).toHaveBeenNthCalledWith(2, 'first');
        });

        it('should support importing arbitrary values using a fallback prefix', async () => {
            const prefixes: ScriptPrefix[] = [
                {
                    prefix: 'ðŸ”º',
                    language: 'javascript',
                },
                {
                    prefix: '#',
                    language: 'text',
                    isFallback: true,
                },
            ];
            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ”ºimport abc from "#other"; globalThis.func1(abc);`,
                }),
                bot2: createPrecalculatedBot('bot2', {
                    other: `test`,
                }),
            };

            const bundle = await bundler.bundleTag(state, `ðŸ”ºmain`, prefixes);

            expect(bundle).not.toEqual(null);
            expect(bundle.source).toBeTruthy();

            eval(bundle.source);

            expect(func1).toBeCalledTimes(1);
            expect(func1).toBeCalledWith('test');
        });
    });

    describe('addLibrary()', () => {
        let func1: jest.Mock<any>;
        let func2: jest.Mock<any>;

        beforeEach(() => {
            require('axios').__reset();
            bundler = new ESBuildPortalBundler();
            (<any>globalThis).func1 = func1 = jest.fn();
            (<any>globalThis).func2 = func2 = jest.fn();
        });

        afterEach(() => {
            delete (<any>globalThis).func2;
            delete (<any>globalThis).func1;
        });

        it('should use the given source when importing the specified library', async () => {
            bundler.addLibrary({
                id: 'casualos',
                source: 'globalThis.func1("casualos");',
                language: 'javascript',
            });

            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ“–import "casualos"; globalThis.func1("main");`,
                }),
            };
            const bundle = await bundler.bundleTag(state, 'main', [
                {
                    prefix: 'ðŸ“–',
                    language: 'typescript',
                },
            ]);

            expect(bundle).toMatchSnapshot();
            expect(bundle.source).toBeTruthy();

            eval(bundle.source);

            expect(func1).toHaveBeenCalledTimes(2);
            expect(func1).toHaveBeenNthCalledWith(1, 'casualos');
            expect(func1).toHaveBeenNthCalledWith(2, 'main');
        });

        it('should be able to process imports from libraries', async () => {
            require('axios').__setResponse({
                data: `export const fun = globalThis.func1;`,
            });

            bundler.addLibrary({
                id: 'casualos',
                source: 'export * from "lodash";',
                language: 'javascript',
            });

            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ“–import { fun } from "casualos"; fun("main");`,
                }),
            };
            const bundle = await bundler.bundleTag(state, 'main', [
                {
                    prefix: 'ðŸ“–',
                    language: 'typescript',
                },
            ]);

            expect(bundle).toMatchSnapshot();
            expect(bundle.source).toBeTruthy();

            eval(bundle.source);

            expect(func1).toHaveBeenCalledTimes(1);
            expect(func1).toHaveBeenNthCalledWith(1, 'main');
        });

        it('should only override imports that exactly match the library name', async () => {
            bundler.addLibrary({
                id: 'rxjs',
                source: 'globalThis.func1("rxjs");',
                language: 'javascript',
            });

            bundler.addLibrary({
                id: 'rxjs/operators',
                source: 'globalThis.func1("rxjs/operators");',
                language: 'javascript',
            });

            const state = {
                bot1: createPrecalculatedBot('bot1', {
                    main: `ðŸ“–import "rxjs/operators"; globalThis.func1("main");`,
                }),
            };
            const bundle = await bundler.bundleTag(state, 'main', [
                {
                    prefix: 'ðŸ“–',
                    language: 'typescript',
                },
            ]);

            expect(bundle).toMatchSnapshot();
            expect(bundle.source).toBeTruthy();

            eval(bundle.source);

            expect(func1).toHaveBeenCalledTimes(2);
            expect(func1).toHaveBeenNthCalledWith(1, 'rxjs/operators');
            expect(func1).toHaveBeenNthCalledWith(2, 'main');
        });
    });
});
