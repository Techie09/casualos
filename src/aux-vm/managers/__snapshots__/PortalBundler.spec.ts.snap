// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ESBuildPortalBundler addLibrary() should be able to process imports from libraries 1`] = `
Object {
  "externals": Object {
    "lodash": Object {
      "id": "lodash",
      "typescriptDefinitionsURL": null,
      "url": "https://cdn.skypack.dev/lodash?dts",
    },
  },
  "libraries": Object {
    "casualos": Object {
      "id": "casualos",
      "language": "javascript",
      "source": "export * from \\"lodash\\";",
    },
  },
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "source": "(() => {
  // http-ns:https://cdn.skypack.dev/lodash?dts
  var fun = globalThis.func1;

  // aux-ns:ðŸ“–bot1.main?auxmodule
  fun(\\"main\\");
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler addLibrary() should only override imports that exactly match the library name 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {
    "rxjs/operators": Object {
      "id": "rxjs/operators",
      "language": "javascript",
      "source": "globalThis.func1(\\"rxjs/operators\\");",
    },
  },
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "source": "(() => {
  // lib-ns:rxjs/operators
  globalThis.func1(\\"rxjs/operators\\");

  // aux-ns:ðŸ“–bot1.main?auxmodule
  globalThis.func1(\\"main\\");
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler addLibrary() should use the given source when importing the specified library 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {
    "casualos": Object {
      "id": "casualos",
      "language": "javascript",
      "source": "globalThis.func1(\\"casualos\\");",
    },
  },
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "source": "(() => {
  // lib-ns:casualos
  globalThis.func1(\\"casualos\\");

  // aux-ns:ðŸ“–bot1.main?auxmodule
  globalThis.func1(\\"main\\");
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() library emoji imports should report errors that occur while fetching data 1`] = `
"Error: Build failed with 1 error:
aux-ns:ðŸ“–bot1.main?auxmodule:1:20: error: [casualos] Error: Get failed."
`;

exports[`ESBuildPortalBundler bundleTag() library emoji should resolve with a bundle that contains the specified tags 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
    "bot2": Set {
      "main",
    },
  },
  "source": "(() => {
  // aux-ns:ðŸ“–bot1.main?auxmodule
  console.log(\\"abc\\");

  // aux-ns:ðŸ“–bot2.main?auxmodule
  console.log(\\"def\\");
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() library emoji should resolve with build errors 1`] = `
Object {
  "error": "Error: Build failed with 1 error:
aux-ns:ðŸ“–bot1.main?auxmodule:1:15: error: Unterminated string literal",
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() multiple emoji imports should report errors that occur while fetching data 1`] = `
"Error: Build failed with 1 error:
aux-ns:ðŸ™‚bot1.main?auxmodule:1:20: error: [casualos] Error: Get failed."
`;

exports[`ESBuildPortalBundler bundleTag() multiple emoji should resolve with a bundle that contains the specified tags 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
    "bot2": Set {
      "main",
    },
  },
  "source": "(() => {
  // aux-ns:ðŸ™‚bot1.main?auxmodule
  console.log(\\"abc\\");

  // aux-ns:ðŸ“–bot2.main?auxmodule
  console.log(\\"def\\");
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() multiple emoji should resolve with build errors 1`] = `
Object {
  "error": "Error: Build failed with 1 error:
aux-ns:ðŸ™‚bot1.main?auxmodule:1:15: error: Unterminated string literal",
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() should support JSON 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "source": "(() => {
})();
",
  "tag": "main",
  "warnings": Array [
    "Ignoring this import because \\"aux-ns:ðŸ“–bot1.main?auxmodule\\" was marked as having no side effects",
  ],
}
`;

exports[`ESBuildPortalBundler bundleTag() should support JSX 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "source": "(() => {
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() should support TSX 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "source": "(() => {
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() should support typescript 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "source": "(() => {
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() smile emoji imports should report errors that occur while fetching data 1`] = `
"Error: Build failed with 1 error:
aux-ns:ðŸ™‚bot1.main?auxmodule:1:20: error: [casualos] Error: Get failed."
`;

exports[`ESBuildPortalBundler bundleTag() smile emoji should resolve with a bundle that contains the specified tags 1`] = `
Object {
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
    "bot2": Set {
      "main",
    },
  },
  "source": "(() => {
  // aux-ns:ðŸ™‚bot1.main?auxmodule
  console.log(\\"abc\\");

  // aux-ns:ðŸ™‚bot2.main?auxmodule
  console.log(\\"def\\");
})();
",
  "tag": "main",
  "warnings": Array [],
}
`;

exports[`ESBuildPortalBundler bundleTag() smile emoji should resolve with build errors 1`] = `
Object {
  "error": "Error: Build failed with 1 error:
aux-ns:ðŸ™‚bot1.main?auxmodule:1:15: error: Unterminated string literal",
  "externals": Object {},
  "libraries": Object {},
  "modules": Object {
    "bot1": Set {
      "main",
    },
  },
  "tag": "main",
  "warnings": Array [],
}
`;
